crates/astria-conductor/src/data_availability.rs

```diff
@@ -0,0 +1,382 @@
+                });
+            }
+        }
+    }
+
+    #[instrument(skip(self, sequencer_data_res))]
+    fn process_sequencer_datas(
+        &mut self,
+        height: u64,
+        sequencer_data_res: Result<
+            eyre::Result<Vec<SignedNamespaceData<SequencerNamespaceData>>>,
+            JoinError,
+        >,
+    ) {
+        let sequencer_data = match sequencer_data_res {
+            Err(e) => {
+                warn!(height, error.message = %e, error.cause = ?e, "task querying celestia for sequencer data failed");
+                return;
+            }
+
+            Ok(Err(e)) => {
+                warn!(error.message = %e, error.cause = ?e, "task querying celestia for sequencer data returned with an error");
+                return;
+            }
+
+            Ok(Ok(sequencer_data)) => sequencer_data,
+        };
+
+        // Set the current block height to the maximum height seen. Having reached this
+        // handler means that we have successfully received valid (but unverified) sequencer
+        // data at celestia `height`. If the next steps fail that is fine: re-requesting
+        // the data will not change the verification failure.
+        // If there are other tasks querying celestia for lower heights are still in
+        // flight they are unaffected and will still be processed here.
+        self.current_block_height = std::cmp::max(self.current_block_height, height);
+        if self.process_sequencer_datas.contains_key(&height) {
+            error!(
+                "sequencer data is already being processed; no two sequencer data responses \
+                 should have been received; this is a bug"
+            );
+            return;
+        }
+        self.process_sequencer_datas.spawn(
+            height,
+            process_sequencer_data(
+                height,
+                sequencer_data,
+                self.celestia_client.clone(),
+                self.block_verifier.clone(),
+                self.namespace,
+            )
+            .in_current_span(),
+        );
+    }
+
+    #[instrument(skip_all, fields(height))]
+    fn send_sequencer_subsets(
+        &self,
+        sequencer_subsets_res: Result<eyre::Result<Vec<SequencerBlockSubset>>, JoinError>,
+    ) -> eyre::Result<()> {
+        let subsets = match sequencer_subsets_res {
+            Err(e) => {
+                warn!(error.message = %e, error.cause = ?e, "task processing sequencer data failed");
+                return Ok(());
+            }
+            Ok(Err(e)) => {
+                warn!(error.message = %e, error.cause = ?e, "task processing sequencer data returned with an error");
+                return Ok(());
+            }
+            Ok(Ok(subsets)) => subsets,
+        };
+        self.executor_tx
+            .send(executor::ExecutorCommand::FromCelestia(subsets))
+            .wrap_err("failed sending processed sequencer subsets: executor channel is closed")
```

Comment:
I believe we have a problematic race condition bug here in 2 different instances:

1. If blocks are received out of order from DA, they will be executed out of order potentially.
2. On startup, if there are blocks being synced (in sequencer_reader loop), and this fires off some new data, again executing out of order.

The first one we can call a bug and hope doesn't happen, create an issue around the second I believe is pretty problematic.

Reply:
Ad 1: yeah, we can fix that by enforcing heights being contiguous. Follow-up?

Ad 2: I have pushed a commit that tries to address this: `sequencer::Reader` gets a oneshot channel over which it notifies conductor if sync is complete. Only once sync is complete will conductor execute `data_availability::Reader::run_until_stopped`. Do you thinks this addresses the issue?

Reply:
Created an issue